crates/wit/src/commands/pull.rs

```diff
@@ -0,0 +1,273 @@
+                config.merge_config(file_config);
+            }
+            config.to_client()
+        };
+
+        let mut pulled = HashSet::new();
+        for pkg in packages {
+            let name = &pkg.name;
+            if !pulled.insert(name.clone()) {
+                continue;
+            }
+            terminal.status("Resolving", format!("package {pkg}"))?;
+
+            match self.pull(&mut client, &pkg).await? {
+                Some((release, decoded)) => {
+                    let root_pkg = &decoded.resolve().packages[decoded.package()];
+
+                    let release_pkg = PackageName {
+                        namespace: name.namespace().to_string(),
+                        name: name.name().to_string(),
+                        version: Some(release.version.clone()),
+                    };
+                    terminal.status("Downloaded", format!("release {release_pkg}"))?;
+
+                    let version = root_pkg.name.version.clone();
+                    if let Some(wit_version) = &version {
+                        let release_version = &release.version;
+                        if wit_version != release_version {
+                            terminal.warn(format!("Release version {release_version} doesn't match WIT package version {wit_version}"))?;
+                        }
+                    }
+
+                    for (package_id, package) in &decoded.resolve().packages {
+                        let name = &package.name;
+                        if !existing_pkgs.insert(name.clone()) {
+                            continue;
+                        }
+                        let path = self.write_package(decoded.resolve(), package_id)?;
+                        terminal.status(
+                            "Wrote",
+                            format!("package {name} to '{path}'", path = path.display()),
+                        )?;
+                    }
+                }
+                None => {
+                    terminal.warn(format!("No package found for {pkg}"))?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn parse_packages_state(&self) -> Result<(BTreeSet<PackageName>, BTreeSet<PackageName>)> {
+        let mut resolve = Resolve::new();
+
+        // If Resolve::push_dir succeeds there are no unresolved deps
+        match resolve.push_dir(&self.wit_dir) {
+            Ok(_) => {
+                return Ok((
+                    resolve.package_names.into_keys().collect(),
+                    Default::default(),
+                ))
+            }
+            Err(err) => log::debug!("Couldn't resolve packages: {err:#}"),
+        }
+
+        // Approximate the Resolve::push_dir process
+        // TODO: Try to expose this logic from wit-parser instead
+        let mut pkgs = match UnresolvedPackage::parse_dir(&self.wit_dir) {
+            Ok(pkg) => BTreeSet::from([pkg.name]),
```

Comment:
The `pkg`'s foreign deps should be added to `deps` here. Otherwise, we only fetch dependencies of dependencies and not the main package's direct dependencies.

Reply:
Ah yeah I lost track of that while refactoring :facepalm:

Reply:
Fixed and refactored a lot of this logic into a `struct PackagesState` like I should have started with :roll_eyes::  546b092